Improve the configuration of excluded User Agents

mnapoli opened the pull request:
Fixes #7978, replaces #8050 

Problems:
- _Excluded User Agents_ is a comma-separated list, but user agents can contain commas
- we want to let users wrap user agents in quotes (see https://github.com/piwik/piwik/issues/7978#issuecomment-121509999) which is currently not supported
- the inline help for _Excluded User Agents_ doesn't say how to format user agent strings (#7978)
- other fields in the page ask for one item per line:
  - it's not consistent with _Excluded User Agents_ (which is comma-separated), so it's confusing
  - in the code it's actually comma-separated, so it's clearly misleading and broken for some users

These problems apply to global and site-specific settings.

Goals:
- [x] consistent formatting for all fields (and both global and site-specific settings)
  - [x] excluded user agents
  - [x] excluded query parameters
  - [x] excluded IPs
- [x] it needs to be "one per line" because of user agents (that can contain commas)
- [x] enclosing quotes need to be accepted (and trimmed) for user agents
- [x] field helps need to match the actual behavior
- [x] migration script
- [x] tests
- [x] update changelog regarding API changes

Notes:
- enclosing quotes for user agents are trimmed (for the other fields I didn't think it would make sense)
- the help text for user agents has also been updated (only the english version), it now explicitly says that user agents should be separated by new lines. What is the best course of action to ensure translations are updated? New translation key?
- The following API methods are now returning an array of strings instead of a list encoded in CSV:
  - `SitesManager.getExcludedIpsGlobal`
  - `SitesManager.getExcludedQueryParametersGlobal`
  - `SitesManager.getExcludedUserAgentsGlobal`
- The following API methods now accept values separated by new lines instead of commas:
  - `SitesManager.setGlobalExcludedIps`
  - `SitesManager.setGlobalExcludedQueryParameters`
  - `SitesManager.setGlobalExcludedUserAgents`

saleemkce committed:
#7978 - fix for user agent quotes issue

mnapoli committed:
Merge branch 'bugfix-user-agent-quotes-issue' of https://github.com/saleemkce/piwik into bug/7978

mnapoli committed:
#7978 Improve the configuration of excluded User Agents

Before this commit, users can configure excluded user agents by separating them with commas. Enclosing quotes are not trimmed.

With this commit, users can configure excluded user agents one per line. Enclosing quotes are trimmed. This is meant to be simpler and more consistent. It also allows to exclude user agents containing both commas and quotes. This change applies both to global settings as well as site-specific settings.

This commit also changes the help text, which now explicitly says that user agents should be separated by new lines.

A migration script is included: it will replace commas by new lines.

Related: #7978, #8050

mnapoli committed:
#8448 Users can now configure excluded query parameters using line returns

Before this commit, users can configure excluded query parameters by separating them with commas.

With this commit, users can configure them one per line. This is meant to be simpler and more consistent (see #8448). This is now also consistent with form inline helps. This change applies both to global settings as well as site-specific settings.

A migration script is included: it will replace commas by new lines.

Related: #7978, #8050

tsteur commented on core/Updates/2.15.0-b1.php:
```diff
@@ -0,0 +1,49 @@
+        // Excluded query parameters
+        $globalExcludedQueryParameters = Option::get(API::OPTION_EXCLUDED_QUERY_PARAMETERS_GLOBAL);
+        $globalExcludedQueryParameters = str_replace(',', "\n", $globalExcludedQueryParameters);
+        Option::set(API::OPTION_EXCLUDED_QUERY_PARAMETERS_GLOBAL, $globalExcludedQueryParameters);
```
FYI: We should write raw SQL statements here instead of using `Option::set`. If we ever change something with the `option` table or `Option` API this update might fail as it will have an "old" structure

mnapoli replied:
done

tsteur commented on core/Updates/2.15.0-b1.php:
```diff
@@ -0,0 +1,49 @@
+        foreach ($sites as $site) {
+            $site['excluded_parameters'] = str_replace(',', "\n", $site['excluded_parameters']);
+            $site['excluded_user_agents'] = str_replace(',', "\n", $site['excluded_user_agents']);
+            $model->updateSite($site, $site['idsite']);
```
Also here we would ideally write raw SQL update statements. Advantage is also that it will be visible in the Updater and everyone will see which SQL statements need to be executed (in case it is done manually)

mnapoli replied:
Good point, will do.

mnapoli replied:
done

mnapoli committed:
Make the update script for #8448 use raw SQL queries

See https://github.com/piwik/piwik/pull/8448#discussion_r35753623 and https://github.com/piwik/piwik/pull/8448#discussion_r35753688

mnapoli committed:
Changelog update for #8448

* The following API methods are now returning an array of strings instead of a list encoded in CSV:
    * `SitesManager.getExcludedQueryParametersGlobal`
    * `SitesManager.getExcludedUserAgentsGlobal`

mnapoli committed:
Fix SitesManager API tests

mnapoli committed:
#8448 Users can now configure excluded IPs using line returns

Before this commit, users can configure excluded IPs by separating them with commas.

With this commit, users can configure them one per line. This is meant to be simpler and more consistent (see #8448). This is now also consistent with form inline helps. This change applies both to global settings as well as site-specific settings.

A migration script is included: it will replace commas by new lines.

Related: #7978, #8050

mnapoli committed:
Update the changelog and documentation for #8448

mnapoli committed:
Merge branch 'master' into bug/7978

# Conflicts:
#	core/Version.php

mnapoli committed:
Fix test and update script

mnapoli committed:
Fix tests

mnapoli commented:
Replaced by #8482
